src/quic/preferredaddress.cc

```
@@ -0,0 +1,147 @@
+                        address.host,
+                        std::to_string(address.port).c_str(),
+                        &hints) == 0 && req->addrinfo != nullptr;
+}
+}  // namespace
+
+Maybe<PreferredAddress::Policy> PreferredAddress::GetPolicy(
+    Environment* env,
+    Local<Value> value) {
+  CHECK(value->IsUint32());
+  uint32_t val = 0;
+  if (value->Uint32Value(env->context()).To(&val)) {
+    switch (val) {
+      case QUIC_PREFERRED_ADDRESS_USE: return Just(Policy::USE);
+      case QUIC_PREFERRED_ADDRESS_IGNORE: return Just(Policy::IGNORE);
+    }
+  }
+  THROW_ERR_INVALID_ARG_VALUE(env,
+      "%d is not a valid preferred address policy", val);
+  return Nothing<Policy>();
+}
+
+PreferredAddress::PreferredAddress(ngtcp2_path* dest,
+                                   const ngtcp2_preferred_addr* paddr)
+    : dest_(dest), paddr_(paddr) {
+  DCHECK_NOT_NULL(paddr);
+  DCHECK_NOT_NULL(dest);
+}
+
+std::optional<const PreferredAddress::AddressInfo>
+PreferredAddress::ipv4() const {
+  return get_address_info<AF_INET>(*paddr_);
+}
+
+std::optional<const PreferredAddress::AddressInfo>
+PreferredAddress::ipv6() const {
+  return get_address_info<AF_INET6>(*paddr_);
+}
+
+void PreferredAddress::Use(const AddressInfo& address) {
+  uv_getaddrinfo_t req;
+  auto on_exit = OnScopeLeave([&] {
+    if (req.addrinfo != nullptr) uv_freeaddrinfo(req.addrinfo);
+  });
+
+  if (resolve(address, &req)) {
+    CHECK_NOT_NULL(req.addrinfo);
+    dest_->remote.addrlen = req.addrinfo->ai_addrlen;
+    memcpy(dest_->remote.addr, req.addrinfo->ai_addr, req.addrinfo->ai_addrlen);
+  }
+}
+
+void PreferredAddress::Set(
+    ngtcp2_transport_params* params,
+    const sockaddr* addr) {
+  CHECK_NOT_NULL(params);
+  CHECK_NOT_NULL(addr);
+  switch (addr->sa_family) {
+    case AF_INET: return copy_to_transport_params<AF_INET>(params, addr);
+    case AF_INET6: return copy_to_transport_params<AF_INET6>(params, addr);
+  }
+  // Any other value is just ignored.
```

Can you add unreachable to here?

---

See the comment in preferredaddress.h, specifically the last sentence.

```
  // The address family (ipv4 or ipv6) will be automatically
  // detected from the given addr. Any other address family
  // will be ignored.
```

If someone just happens to specify some other address family, we're just going to silently ignore it at this level.